plugins/oauth/girder_oauth/providers/microsoft.py

```diff
@@ -0,0 +1,86 @@
+# -*- coding: utf-8 -*-
+from msal import ConfidentialClientApplication
+
+from girder.exceptions import RestException
+from girder.models.setting import Setting
+
+from ..settings import PluginSettings
+from .base import ProviderBase
+
+
+class Microsoft(ProviderBase):
+    _AUTH_SCOPES = ['User.Read']
+    _API_USER_URL = 'https://graph.microsoft.com/v1.0/me'
+
+    @classmethod
+    def _authority(cls):
+        tenantId = Setting().get(PluginSettings.MICROSOFT_TENANT_ID) or 'common'
+        return f'https://login.microsoftonline.com/{tenantId}'
+
+    def getClientIdSetting(self):
+        return Setting().get(PluginSettings.MICROSOFT_CLIENT_ID)
+
+    def getClientSecretSetting(self):
+        return Setting().get(PluginSettings.MICROSOFT_CLIENT_SECRET)
+
+    @classmethod
+    def getUrl(cls, state):
+        clientId = Setting().get(PluginSettings.MICROSOFT_CLIENT_ID)
+        if not clientId:
+            raise Exception('No Microsoft client ID setting is present.')
+
+        clientSecret = Setting().get(PluginSettings.MICROSOFT_CLIENT_SECRET)
+        if not clientSecret:
+            raise Exception('No Microsoft client secret setting is present.')
+
+        app = ConfidentialClientApplication(
+            client_id=clientId,
+            client_credential=clientSecret,
+            authority=cls._authority(),
+        )
+        # The default response type is 'code', so we don't need to pass it
+        url = app.get_authorization_request_url(
+            scopes=cls._AUTH_SCOPES,
+            state=state,
```

Comment:
Doesn't this call need to pass a callback url?  `redirect_url: '/'.join((getApiUrl(), 'oauth', 'microsoft', 'callback'))`?

Reply:
I don't think so, it is already stored in the app registration: https://docs.microsoft.com/en-us/azure/active-directory/develop/quickstart-register-app#add-a-redirect-uri

It's been working so far for us.